rotkehlchen/tasks/manager.py

```diff
@@ -538,6 +541,36 @@ def _maybe_update_snapshot_balances(self) -> Optional[list[gevent.Greenlet]]:
         """
         with self.database.conn.read_ctx() as cursor:
             if self.database.should_save_balances(cursor):
+                detect_tokens: dict[SupportedBlockchain, list[ChecksumEvmAddress]] = defaultdict(list)  # noqa: E501
+                if CachedSettings().get_settings().auto_detect_tokens is True:
+                    type_filter = [event_type.serialize() for event_type in [
+                        HistoryEventType.RECEIVE,
+                        HistoryEventType.WITHDRAWAL,
+                    ]]
+                    subtype_filter = [event_subtype.serialize() for event_subtype in [
+                        HistoryEventSubType.RECEIVE,
+                        HistoryEventSubType.REWARD,
+                        HistoryEventSubType.REMOVE_ASSET,
+                        HistoryEventSubType.MEV_REWARD,
+                    ]]
+                    last_save_time = self.database.get_last_balance_save_time(cursor)
+                    for chain in EVM_CHAIN_IDS_WITH_TRANSACTIONS:
+                        if len(location_labels := cursor.execute(  # number of earliest receive events of distinct assets after last_save_time  # noqa: E501
+                            'SELECT DISTINCT location_label FROM '
+                            '(SELECT MIN(timestamp), location_label FROM history_events '
```

Comment:
I was thinking that if we have many addresses this task can be really slow. Not sure if it is worth to split the job per chain but it complicates the logic

Reply:
also we query different (ether)scan APIs for each chain